README.md
# Crockford32
A fast little-endian implementation of [Douglas Crockford's Base32 specification](https://www.crockford.com/base32.html).

[![Gem Version](https://badge.fury.io/rb/crockford32.svg)](https://badge.fury.io/rb/crockford32)

## What's in the box?
✅ Simple usage documentation written to get started fast. [Check it out!](#usage)

⚡ A pretty fast implementation of Crockford32 in pure ruby. [Check it out!](#benchmarks)

📚 YARD generated API documentation for the library. [Check it out!](https://tinychameleon.github.io/crockford32/)

🤖 RBS types for your type checking wants. [Check it out!](./sig/crockford32.rbs)

💎 Tests against many Ruby versions. [Check it out!](#ruby-versions)

🔒 MFA protection on all gem owners. [Check it out!](https://rubygems.org/gems/crockford32)


## Installation
Add this line to your application's Gemfile:

```ruby
gem 'crockford32'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install crockford32

### Ruby Versions
This gem is tested against the following Ruby versions:

- 2.6.0
- 2.6.9
- 2.7.0
- 2.7.5
- 3.0.0
- 3.0.3
- 3.1.0
- 3.1.1


## Usage
Encode data with the `encode` method:

```ruby
encoded = Crockford32.encode(1234) # encoded = "J61"
```
See the [the encode API documentation for error details](https://tinychameleon.github.io/crockford32/Crockford32.html#encode-class_method).

---

Decode a value with the `decode` method:

```ruby
decoded = Crockford32.decode("J61") # decoded = 1234
```
See the [the decode API documentation for error details](https://tinychameleon.github.io/crockford32/Crockford32.html#decode-class_method).


### Strings
You can also encode and decode strings by providing the `:string` value as the `into:` argument to `decode`.
Encoding a string requires no special consideration.

```ruby
encoded = Crockford32.encode("abc") # encoded = "1KR66"
decoded = Crockford32.decode(encoded, into: :string) # decoded = "abc"
```

When decoding you may receive a byte string if the encoded value is not alphanumeric:

```ruby
decoded = Crockford32.decode("J61", into: :string) # decoded = "\xD2\x04"
```

### Padding to a Length
You can ensure your encoded values are a specific length by specifying a `length:` on `encode`:

```ruby
encoded = Crockford32.encode(1234, length: 5) # encoded = "J6100"
```

The padding is always "0" values and is always appended to the end of the encoded string.
Whatever you specify as the `length:` will be the length of the encoded string you receive.

Decoding also supports a `length:` argument for symmetry when using `into: :string`:

```ruby
decoded = Crockford32.decode("J61", length: 5, into: :string) # decoded = "\xD2\x04\x00\x00\x00"
```

Any length given to `Crockford32.decode` will not truncate valid decoding results. The `length:`
is treated as a minimum.

### Check Symbols
If you wish to append a check symbol for simple modulus-based error detection both `encode` and `decode` support it with the `check:` keyword.

```ruby
encoded = Crockford32.encode(1234, check: true) # encoded = "J61D". "D" is the checksum symbol.
decoded = Crockford32.decode(encoded, check: true) # decoded = 1234. Checksum is tested.
```

An error will be raised if the decoding process does not pass the checksum:

```ruby
decoded = Crockford32.decode("J71D", check: true) # Notice the 6 changed to a 7.
~/.../crockford32/lib/crockford32.rb:43:in `decode': Value J71 has checksum 8 but requires 13 (Crockford32::ChecksumError)
	from (irb):2:in `<main>'
	from bin/console:8:in `<main>'
```

If you specify a `length:` and `check: true` when encoding, the checksum will be included as part of the `length:`.

### Encoding Type Support
This library currently supports encoding `Integer` and `String` values.

### Little-Endian Encoding
The encoding example above shows the little-endian results of encoding a number.
Some libraries encode using big-endian and will return `"16J"` when the number `1234` is encoded.

This library _always_ uses little-endian.

### More Information
For more detailed information about the library [see the API documentation](https://tinychameleon.github.io/crockford32/).


## Contributing

### Development
To get started development on this gem run the `bin/setup` command. This will install dependencies and run the tests and linting tasks to ensure everything is working.

For an interactive console with the gem loaded run `bin/console`.


### Testing
Use the `bundle exec rake test` command to run unit tests. To install the gem onto your local machine for general integration testing use `bundle exec rake install`.

To test the gem against each supported version of Ruby use `bin/test_versions`. This will create a Docker image for each version and run the tests and linting steps.


### Releasing
Do the following to release a new version of this gem:

- Update the version number in [lib/crockford32/version.rb](./lib/crockford32/version.rb)
- Ensure necessary documentation changes are complete
- Ensure changes are in the [CHANGELOG.md](./CHANGELOG.md)
- Create the new release using `bundle exec rake release`

After this is done the following side-effects should be visible:

- A new git tag for the version number should exist
- Commits for the new version should be pushed to GitHub
- The new gem should be available on [rubygems.org](https://rubygems.org).

Finally, update the documentation hosted on GitHub Pages:

- Check-out the `gh-pages` branch
- Merge `main` into the `gh-pages` branch
- Generate the documentation with `bundle exec rake yard`
- Commit the documentation on the `gh-pages` branch
- Push the new documentation so GitHub Pages can deploy it


## Benchmarks
Benchmarking is tricky and the goal of a benchmark should be clear before attempting performance improvements. The goal of this library for performance is as follows:

> This library should be capable of encoding and decoding IDs at a rate which does not make it a bottleneck for the majority of web APIs.

Given the above goal statement, these benchmarks run on the following environment:

| Attribute | Value |
|:--|--:|
| Ruby Version | 3.1.0 |
| MacOS Version | Catalina 10.15.7 (19H1615) |
| MacOS Model Identifier | MacBookPro10,1 |
| MacOS Processor Name | Quad-Core Intel Core i7 |
| MacOS Processor Speed | 2.7 GHz |
| MacOS Number of Processors | 1 |
| MacOS Total Number of Cores | 4 |
| MacOS L2 Cache (per Core) | 256 KB |
| MacOS L3 Cache | 6 MB |
| MacOS Hyper-Threading Technology | Enabled |
| MacOS Memory | 16 GB |

When run using a constant 16 bytes of data in the above environment the performance is approximately as follows:

```
~/…/crockford32› date && ruby test/benchmarks/current.rb
Fri Feb 18 15:35:49 PST 2022
Warming up --------------------------------------
              encode    11.498k i/100ms
              decode    10.550k i/100ms
       encode string     7.686k i/100ms
       decode string     6.798k i/100ms
Calculating -------------------------------------
              encode    117.614k (± 3.5%) i/s -    597.896k in   5.090119s
              decode    108.253k (± 3.2%) i/s -    548.600k in   5.073326s
       encode string     78.690k (± 3.1%) i/s -    399.672k in   5.084255s
       decode string     69.707k (± 3.5%) i/s -    353.496k in   5.077700s
```

Being conservative in estimation 30k i/s round trip decoding and encoding cycles should be possible. This achieves the performance goal.


CHANGELOG.md
# Changelog

## v1.1.0 - 2022-03-08
- Add `length:` parameter to `Crockford32.decode` to support padding to a given length after decoding into a string.

## v1.0.2 - 2022-02-22
- Add type definitions

## v1.0.1 - 2022-02-19
- Fixed `TypeError` in `private_class_method` call in Ruby < 3.0.0

## v1.0.0 - 2022-02-19
- Initial release


test/test_errors.rb
# frozen_string_literal: true

require "test_helper"

class TestCrockford32Errors < Minitest::Test
  def test_length_too_small_error
    err = ::Crockford32::LengthTooSmallError.new("ABC", 6, 4)
    assert_kind_of ::Crockford32::EncodeError, err
    assert_equal "Encoding ABC requires a minimum length of 6, but received 4", err.message
  end

  def test_unsupported_encoding_type_error
    err = ::Crockford32::UnsupportedEncodingTypeError.new(Regexp)
    assert_kind_of ::Crockford32::EncodeError, err
    assert_equal "Encoding Regexp not supported", err.message
  end

  def test_invalid_character_error
    err = ::Crockford32::InvalidCharacterError.new("AUBC", 1)
    assert_kind_of ::Crockford32::DecodeError, err
    assert_equal "Invalid character 'U' in 'AUBC' at index 1", err.message
  end

  def test_invalid_target_type_error
    err = ::Crockford32::UnsupportedDecodingTypeError.new(:apple)
    assert_kind_of ::Crockford32::DecodeError, err
    assert_equal "Decoding as :apple not supported", err.message
  end

  def test_checksum_error
    err = ::Crockford32::ChecksumError.new("ABC", 1, 2)
    assert_kind_of ::Crockford32::DecodeError, err
    assert_equal "Value ABC has checksum 1 but requires 2", err.message
  end
end


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "crockford32"

require "debug"
require "minitest/autorun"


test/test_crockford32.rb
# frozen_string_literal: true

require "test_helper"

class TestCrockford32 < Minitest::Test
  def test_it_has_a_version_number
    refute_nil ::Crockford32::VERSION
  end
end


lib/crockford32/errors.rb
# frozen_string_literal: true

module Crockford32
  # The base error for all possible errors from {Crockford32}
  class Error < StandardError; end

  # The base error for all errors from {Crockford32.encode}
  class EncodeError < Error; end

  # The base error for all errors from {Crockford32.decode}
  class DecodeError < Error; end

  # An error representing the length of an encoded value being larger than the
  # requested maximum length.
  class LengthTooSmallError < EncodeError
    # @param value [Integer, String] the value to be encoded.
    # @param needed [Integer] the needed length to encode the value.
    # @param given [Integer] the specified length to encode the value.
    def initialize(value, needed, given)
      super("Encoding #{value} requires a minimum length of #{needed}, but received #{given}")
    end
  end

  # An error representing an attmept to encode an unsupported type.
  class UnsupportedEncodingTypeError < EncodeError
    # @param type [Type] the unsupported type of the encode attempt.
    def initialize(type)
      super("Encoding #{type} not supported")
    end
  end

  # An error representing a decode operation finding an illegal symbol within a
  # provided value.
  class InvalidCharacterError < DecodeError
    # @param string [String] the string to decode.
    # @param index [Integer] the index within the string of the invalid character.
    def initialize(string, index)
      super("Invalid character '#{string[index]}' in '#{string}' at index #{index}")
    end
  end

  # An error representing an unsupported destination type for a decode operation.
  class UnsupportedDecodingTypeError < DecodeError
    # @param type [Symbol] the supplied destination type.
    def initialize(type)
      super("Decoding as :#{type} not supported")
    end
  end

  # An error representing a checksum mismatch between the decoded value and the
  # supplied check symbol.
  class ChecksumError < DecodeError
    # @param value [String] the Base32 value to decode.
    # @param checksum [Integer] the checksum of the decoded value.
    # @param checksum_required [Integer] the required checksum of the decoded value.
    def initialize(value, checksum, checksum_required)
      super("Value #{value} has checksum #{checksum} but requires #{checksum_required}")
    end
  end
end


lib/crockford32.rb
# frozen_string_literal: true

require_relative "crockford32/version"
require_relative "crockford32/errors"

# A fast little-endian implementation of {https://www.crockford.com/base32.html Douglas Crockford’s Base32 specification}.
#
# @since 1.0.0
module Crockford32
  # The number of bits encoded per symbol.
  ENCODED_BITS = 0x05

  # The minimum value of a check symbol.
  CHECK_SYMBOL_MIN_VALUE = 0x20

  # The prime number used to implement error detection.
  CHECKSUM_PRIME = 0x25

  # The ordinal value of an ASCII dash character.
  DASH = "-".ord.freeze

  # standard:disable Layout/ExtraSpacing,Layout/ArrayAlignment

  # Symbol values in order by encoded ASCII ordinal values.
  DECODE_ORDINALS = [
    nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil,
    nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil,
    nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil,
     34, nil, nil, nil, nil, nil,  32, nil, nil, nil, nil, nil,
      0,   1,   2,   3,   4,   5,   6,   7,   8,   9, nil, nil,
    nil,  35, nil, nil, nil,  10,  11,  12,  13,  14,  15,  16,
     17,   1,  18,  19,   1,  20,  21,   0,  22,  23,  24,  25,
     26,  36,  27,  28,  29,  30,  31, nil, nil, nil, nil, nil,
    nil,  10,  11,  12,  13,  14,  15,  16,  17,   1,  18,  19,
      1,  20,  21,   0,  22,  23,  24,  25,  26,  36,  27,  28,
     29,  30,  31, nil, nil, nil,  33
  ].freeze

  # Encoding symbols ordered by bit value.
  ENCODE_SYMBOLS = [
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F",
    "G", "H", "J", "K", "M", "N", "P", "Q", "R", "S", "T", "V", "W", "X", "Y", "Z",
    "*", "~", "$", "=", "U"
  ].freeze

  # standard:enable Layout/ExtraSpacing,Layout/ArrayAlignment

  # @!group Public Methods

  # Decode a Base32 value.
  #
  # @since 1.0.0
  #
  # @param value [String] the Base32 value to decode.
  # @param into [Symbol] the destination type to decode into. Can be +:integer+ or +:string+.
  # @param check [Boolean] whether to validate the check symbol.
  # @param length [Integer, nil] the length of the resulting string when right padded.
  #
  # @return [Integer, String] the decoded value.
  #
  # @raise [ChecksumError] when the check symbol does not match the decoded checksum.
  # @raise [InvalidCharacterError] when the value being decoded has a character outside the
  #   Base32 symbol set or a misplaced check symbol.
  # @raise [UnsupportedDecodingTypeError] when the requested +into:+ type is not supported.
  def self.decode(value, into: :integer, check: false, length: nil)
    checksum = check ? value[-1] : nil
    value = check ? value[0...-1] : value

    result = le_decode_number value

    if check
      actual = result % CHECKSUM_PRIME
      required = DECODE_ORDINALS[checksum.ord]
      raise ChecksumError.new(value, actual, required) if actual != required
    end

    convert result, into, length
  end

  # Encode a value as Base32.
  #
  # @since 1.0.0
  #
  # @param value [Integer, String] the value to encode.
  # @param length [Integer] the exact length of the Base32 string. Will be padded with "0" to meet length.
  # @param check [Boolean] whether to include a check symbol. This symbol is included in the length.
  #
  # @return [String] the encoded value.
  #
  # @raise [LengthTooSmallError] when the requested +length:+ is not large enough to fit the encoded result.
  # @raise [UnsupportedEncodingTypeError] when the value to encode is an unsupported type.
  def self.encode(value, length: nil, check: false)
    le_encode_number(raw_value_to_number(value), length, check)
  end

  # @!endgroup

  # @!group Private Methods
  # @!visibility private

  # Decode a value with the expectation that it is in little-endian order.
  #
  # @param encoded_value [String] the value to decode.
  # @return [Integer] the decoded value.
  def self.le_decode_number(encoded_value)
    symbol = -1
    bits = -ENCODED_BITS
    encoded_value.bytes.reduce(0) do |result, ch|
      symbol += 1
      next result if ch == DASH
      val = DECODE_ORDINALS[ch.ord]
      raise InvalidCharacterError.new(encoded_value, symbol) if val.nil? || val >= CHECK_SYMBOL_MIN_VALUE
      bits += ENCODED_BITS
      result | (val << bits)
    end
  end

  # Convert a decoded result into the destination type.
  #
  # @param result [Integer] the decoded value.
  # @param type [Symbol] the destination type for the value. Can be +:integer+ or +:string+.
  # @param length [Integer, nil] the length to pad the string to.
  # @return [Integer, String] the decoded value converted to the destination type.
  def self.convert(result, type, length)
    case type
    when :integer
      result
    when :string
      into_string result, length
    else
      raise UnsupportedDecodingTypeError.new(type)
    end
  end

  # Convert an Integer into a String.
  #
  # Each 8-bit sequence is packed into a String in little-endian order.
  #
  # @param result [Integer] the decoded value.
  # @param length [Integer, nil] the length of the decoded value with right padding.
  # @return [String] the decoded value as a String.
  def self.into_string(result, length)
    q, r = result.bit_length.divmod(0x08)
    q += 1 if r > 0
    bytes = Array.new(q)
    q.times do |i|
      bytes[i] = result & 0xff
      result >>= 0x08
    end

    bstr = bytes.pack("C*")
    return bstr if length.nil?

    bstr.ljust(length, "\x00")
  end

  # Convert a raw value into an Integer for encoding.
  #
  # @param value [Integer, String] the value being encoded.
  # @return [Integer] the value converte to an Integer.
  def self.raw_value_to_number(value)
    case value
    when String
      q, r = value.bytesize.divmod(8)
      if r == 0
        string_to_integer_unrolled value, q
      else
        string_to_integer value
      end
    when Integer
      value
    else
      raise UnsupportedEncodingTypeError.new value.class
    end
  end

  # Convert a String to an Integer one byte per iteration.
  #
  # @param s [String] the String to convert.
  # @return [Integer] the String converted to an Integer in little-endian order.
  def self.string_to_integer(s)
    shift = -0x08
    s.each_byte.reduce(0) do |n, b|
      shift += 0x08
      n + (b << shift)
    end
  end

  # Convert a String to an Integer 8 bytes per iteration.
  # @param s [String] the String to convert.
  # @param iterations [Integer] the number of iterations to perform.
  # @return [Integer] the String converted to an Integer in little-endian order.
  def self.string_to_integer_unrolled(s, iterations)
    n = 0
    bytes = s.bytes
    while iterations > 0
      o = (iterations - 1) * 0x40
      i = iterations * 0x08
      n += bytes[i - 1] << (o + 0x38)
      n += bytes[i - 2] << (o + 0x30)
      n += bytes[i - 3] << (o + 0x28)
      n += bytes[i - 4] << (o + 0x20)
      n += bytes[i - 5] << (o + 0x18)
      n += bytes[i - 6] << (o + 0x10)
      n += bytes[i - 7] << (o + 0x08)
      n += bytes[i - 8] << (o + 0x00)
      iterations -= 1
    end
    n
  end

  # Encode an Integer as a Base32 value.
  #
  # @see encode
  # @return [String]
  def self.le_encode_number(number, length, check)
    result = +""
    n = number
    loop do
      chunk = n & 0x1F
      result << ENCODE_SYMBOLS[chunk]
      n >>= ENCODED_BITS
      break if n == 0
    end

    rlen = result.length + (check ? 1 : 0)
    if length
      raise LengthTooSmallError.new(number, rlen, length) if rlen > length
      result << "0" * (length - rlen)
    end

    result << ENCODE_SYMBOLS[number % CHECKSUM_PRIME] if check
    result.freeze
  end

  private_class_method :le_decode_number, :convert, :into_string, :raw_value_to_number,
    :string_to_integer, :string_to_integer_unrolled, :le_encode_number

  # @!endgroup
end
